vochain/state_snapshot.go

```
@@ -0,0 +1,419 @@
+//
+// - treeN is the raw bytes dump of all trees.
+type StateSnapshot struct {
+	path               string
+	file               *os.File
+	lock               sync.Mutex
+	header             SnapshotHeader
+	headerSize         uint32
+	currentTree        int // indicates the current tree
+	treeReadPointer    int // points to the last read byte
+	currentTreeReadEnd int // points to the last byte of the current tree
+}
+
+// SnapshotHeader is the header structure of StateSnapshot containing the list of merkle trees.
+type SnapshotHeader struct {
+	Version int
+	Root    []byte
+	ChainID string
+	Height  uint32
+	Trees   []SnapshotHeaderTree
+	// checksum?
+}
+
+// SnapshotHeaderTree represents a merkle tree of the StateSnapshot.
+type SnapshotHeaderTree struct {
+	Name   string
+	Size   uint32
+	Parent string
+	Root   []byte
+}
+
+// SetMainRoot sets the root for the mail merkle tree of the state.
+func (s *StateSnapshot) SetMainRoot(root []byte) {
+	s.header.Root = root
+}
+
+// SetHeight sets the blockchain height for the snapshot.
+func (s *StateSnapshot) SetHeight(height uint32) {
+	s.header.Height = height
+}
+
+// SetChainID sets the blockchain identifier for the snapshot.
+func (s *StateSnapshot) SetChainID(chainID string) {
+	s.header.ChainID = chainID
+}
+
+// Open reads an existing snapshot file and decodes the header.
+// After calling this method everything is ready for reading the first
+// merkle tree. No need to execute `FetchNextTree` until io.EOF is reached.
+//
+// This method performs the opposite operation of `Create`, one of both needs
+// to be called (but not both).
+func (s *StateSnapshot) Open(filePath string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	var err error
+	s.path = filePath
+	s.file, err = os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	headerSizeBytes := make([]byte, 32)
+	if _, err = s.file.Read(headerSizeBytes); err != nil {
+		return err
+	}
```

note that `Read` may read fewer bytes. You need `io.ReadFull` to actually ensure you read 32 bytes. this is a fairly common gotcha as in practice short reads do read fully, but you can still be caught by weird bugs where you silently ignore bytes in edge cases.